rdoff/rdf2bin.1
.TH RDF2BIN 1 "September 6, 1999" "Debian Project" "Debian Manual"
.SH NAME
rdf2bin, rdf2com \- convert an RDOFF object file to flat binary
.SH SYNOPSIS
.B rdf2bin
.RI "[\-o " relocation-origin ]
.RI "[\-p " segment-alignment ]
.RI "[\-f" format ]
.I input-file
.I output-file
.br
.B rdf2com
.RI "[\-p " segment-alignment ]
.I input-file
.I output-file
.br
.B rdf2ith
.RI "[\-o " relocation-origin ]
.RI "[\-p " segment-alignment ]
.I input-file
.I output-file
.br
.B rdf2srec
.RI "[\-o " relocation-origin ]
.RI "[\-p " segment-alignment ]
.I input-file
.I output-file
.SH OPTIONS
.TP
.RI "\-o " relocation-origin
Relocate at origin
.IR relocation-origin .
If invoked as
.BR rdf2com ,
the default relocation origin will be 0x100. Else, the default origin is 0.
.TP
.RI "\-p " segment-alignment
Pad segments until their size is a multiple of
.IR segment-alignment .
By default, 16 is used.
.TP
.RI "\-f " format
Specify the output format.  The currently supported formats are binary
.RI ( bin ),
DOS COM (binary with origin 0x100)
.RI ( com )
Intel hex
.RI ( ith
or
.IR ihx ),
and
Motorola S-Records
.RI ( srec ).
If not specified, the format is set by the command name.
.SH AUTHORS
Julian Hall <jules@earthcorp.com>, H. Peter Anvin <hpa@zytor.com>.
.PP
This manual page was written by Matej Vela <vela@debian.org>.
.SH BUGS
This utility currently only supports the classic segments
.IR .text ,
.I .data
and
.IR .bss .



CHANGES
2.04
----
* Sanitize macro handing in the %error directive.
* New %warning preprocessor directive.
* __utf16__ and __utf32__ operators to generate UTF-16 and UTF-32
  strings.
* Fix bug in case-insensitive matching when compiled on platforms that
  don't use the "configure" script.  Of the official release binaries,
  that only affects the OS/2 binary.
* Correct the handling of nested %reps.
* Support for x87 packed BCD constants.
* New %warning directive to issue user-controlled warnings.
* New %strcat directive to join strings together.
* Correct the LTR instruction in 64-bit mode.
* Fix unnecessary REX.W prefix on indirect jumps in 64-bit mode.

2.03.01
-------
* Fix buffer overflow in the listing module.
* Fix the handling of hexadecimal escape codes in `...` strings.
* The Postscript/PDF documentation has been reformatted.
* The -F option now implies -g.

2.03
----
* Add support for Intel AVX, CLMUL and FMA instructions,
  including YMM registers.
* "dy", "resy" and "yword" for 32-byte operands.
* Fix some SSE5 instructions.
* Intel INVEPT, INVVPID and MOVBE instructions.
* Fix checking for critical expressions when the optimizer is enabled.
* Support the DWARF debugging format for ELF targets.
* Fix optimizations of signed bytes.
* Fix operation on bigendian machines.
* Fix buffer overflow in the preprocessor.
* SAFESEH support for Win32, IMAGEREL for Win64 (SEH).
* %? and %?? to refer to the name of a macro itself.  In particular,
  "%idefine keyword $%?" can be used to make a keyword "disappear".
* New options for dependency generation: -MD, -MF, -MP, -MT, -MQ.
* New preprocessor directives %pathsearch and %depend; INCBIN
  reimplemented as a macro.
* %include now resolves macros in a sane manner.
* %substr can now be used to get other than one-character substrings.
* New type of character/string constants, using backquotes (`...`),
  which support C-style escape sequences.
* %defstr and %idefstr to stringize macro definitions before creation.
* Fix forward references used in EQU statements.

2.02
----
* Additional fixes for MMX operands with explicit "qword", as well as
  (hopefully) SSE operands with "oword".
* Fix handling of truncated strings with DO.
* Fix segfaults due to memory overwrites when floating-point constants
  were used.
* Fix segfaults due to missing include files.
* Fix OpenWatcom Makefiles for DOS and OS/2.
* Add autogenerated instruction list back into the documentation.
* ELF: Fix segfault when generating stabs, and no symbols have been
  defined.
* ELF: Experimental support for DWARF debugging information.
* New compile date and time standard macros.
* %ifnum now returns true for negative numbers.
* New %iftoken test for a single token.
* New %ifempty test for empty expansion.
* Add support for the XSAVE instruction group.
* Makefile for Netware/gcc.
* Fix issue with some warnings getting emitted way too many times.
* Autogenerated instruction list added to the documentation.

2.01
----
* Fix the handling of MMX registers with explicit "qword" tags on
  memory (broken in 2.00 due to 64-bit changes.)
* Fix the PREFETCH instructions.
* Fix the documentation.
* Fix debugging info when using "-f elf" (backwards alias for "-f
  elf32").
* Man pages for rdoff tools (from the Debian project.)
* ELF: handle large numbers of sections.
* Fix corrupt output when the optimizer runs out of passes.

2.00
----
* Added c99 data-type compliance.
* Added general x86-64 support.
* Added win64 (x86-64 COFF) output format.
* Added __BITS__ standard macro.
* Renamed the elf output format to elf32 for clarity.
* Added elf64 and macho (MacOS X) output formats.
* Added Numeric constants in DQ directive.
* Added oword, do and reso pseudo operands.
* Allow underscores in numbers.
* Added 8-, 16- and 128-bit floating-point formats.
* Added binary, octal and hexadecimal floating-point.
* Correct the generation of floating-point constants.
* Added Floating-point option control.
* Added Infinity and NaN floating point support.
* Added ELF Symbol Visibility support.
* Added Setting OSABI value in ELF header directive.
* Added Generate Makefile Dependencies option.
* Added Unlimited Optimization Passes option.
* Added %IFN and %ELIFN support.
* Added Logical Negation Operator.
* Enhanced Stack Relative Preprocessor Directives.
* Enhanced ELF Debug Formats.
* Enhanced Send Errors to a File option.
* Added SSSE3, SSE4.1, SSE4.2, SSE5 support.
* Added a large number of additional instructions.
* Significant performance improvements.

0.98.39
-------
* fix buffer overflow
* fix outas86's .bss handling
* "make spotless" no longer deletes config.h.in.
* %(el)if(n)idn insensitivity to string quotes difference (#809300).
* (nasm.c) __OUTPUT_FORMAT__ changed to string value instead of symbol.

0.98.38
-------
* Add Makefile for 16-bit DOS binaries under OpenWatcom, and modify
  mkdep.pl to be able to generate completely pathless dependencies, as
  required by OpenWatcom wmake (it supports path searches, but not
  explicit paths.)
* Fix the STR instruction.
* Fix the ELF output format, which was broken under certain
  circumstances due to the addition of stabs support.
* Quick-fix Borland format debug-info for -f obj  
* Fix for %rep with no arguments (#560568)
* Fix concatenation of preprocessor function call (#794686)
* Fix long label causes coredump (#677841)
* Use autoheader as well as autoconf to keep configure from generating
  ridiculously long command lines.
* Make sure that all of the formats which support debugging output
  actually will suppress debugging output when -g not specified.

0.98.37
-------
* Paths given in "-I" switch searched for "incbin"ed as
  well as "%include"ed files.
* Added stabs debugging for the ELF output format, patch from
  Martin Wawro.
* Fix output/outbin.c to allow origin > 80000000h.
* Make -U switch work.
* Fix the use of relative offsets with explicit prefixes, e.g. "a32 loop foo".
* Remove "backslash()".
* Fix the SMSW and SLDT instructions.

rdoff/rdf2bin.c
/* ----------------------------------------------------------------------- *
 *   
 *   Copyright 1996-2009 The NASM Authors - All Rights Reserved
 *   See the file AUTHORS included with the NASM distribution for
 *   the specific copyright holders.
 *
 *   Redistribution and use in source and binary forms, with or without
 *   modification, are permitted provided that the following
 *   conditions are met:
 *
 *   * Redistributions of source code must retain the above copyright
 *     notice, this list of conditions and the following disclaimer.
 *   * Redistributions in binary form must reproduce the above
 *     copyright notice, this list of conditions and the following
 *     disclaimer in the documentation and/or other materials provided
 *     with the distribution.
 *     
 *     THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND
 *     CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
 *     INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
 *     MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 *     DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
 *     CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 *     SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
 *     NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 *     LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 *     HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 *     CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
 *     OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE,
 *     EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *
 * ----------------------------------------------------------------------- */

/*
 * rdf2bin.c - convert an RDOFF object file to flat binary
 */

#include "compiler.h"

#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <ctype.h>
#include <errno.h>

#include "rdfload.h"
#include "nasmlib.h"

const char *progname;

static uint32_t origin = 0;
static bool origin_def = false;
static uint32_t align = 16;
static bool align_def = false;

struct output_format {
    const char *name;
    const char *mode;
    int (*init)(FILE *f);
    int (*output)(FILE *f, void *data, uint32_t bytes, uint32_t where);
    int (*fini)(FILE *f);
};

static int null_init_fini(FILE *f)
{
    (void)f;
    return 0;
}

static int com_init(FILE *f)
{
    (void)f;
    if (!origin_def)
	origin = 0x100;
    return 0;
}

static int output_bin(FILE *f, void *data, uint32_t bytes, uint32_t where)
{
    static uint32_t offset = 0;	/* Current file offset, if applicable */
    size_t pad;

    if (where-origin < offset) {
	fprintf(stderr, "%s: internal error: backwards movement\n", progname);
	exit(1);
    }

    pad = (where-origin) - offset;
    if (fwritezero(pad, f) != pad)
	return -1;
    offset += pad;

    if (fwrite(data, 1, bytes, f) != bytes)
	return -1;
    offset += bytes;

    return 0;
}

static int write_ith_record(FILE *f, unsigned int len, uint16_t addr,
			    uint8_t type, void *data)
{
    char buf[1+2+4+2+255*2+2+2];
    char *p = buf;
    uint8_t csum, *dptr = data;
    unsigned int i;

    if (len > 255) {
	fprintf(stderr, "%s: internal error: invalid ith record size\n",
		progname);
	exit(1);
    }

    csum = len + addr + (addr >> 8) + type;
    for (i = 0; i < len; i++)
	csum += dptr[i];
    csum = -csum;

    p += sprintf(p, ":%02X%04X%02X", len, addr, type);
    for (i = 0; i < len; i++)
	p += sprintf(p, "%02X", dptr[i]);
    p += sprintf(p, "%02X\n", csum);

    if (fwrite(buf, 1, p-buf, f) != (size_t)(p-buf))
	return -1;

    return 0;
}

static int output_ith(FILE *f, void *data, uint32_t bytes, uint32_t where)
{
    static uint32_t last = 0;	/* Last address written */
    uint8_t abuf[2];
    uint8_t *dbuf = data;
    uint32_t chunk;

    while (bytes) {
	if ((where ^ last) & ~0xffff) {
	    abuf[0] = where >> 24;
	    abuf[1] = where >> 16;
	    if (write_ith_record(f, 2, 0, 4, abuf))
		return -1;
	}

	/* Output up to 32 bytes, but always end on an aligned boundary */
	chunk = 32 - (where & 31);
	if (bytes < chunk)
	    chunk = bytes;

	if (write_ith_record(f, chunk, (uint16_t)where, 0, dbuf))
	    return -1;

	dbuf += chunk;
	last = where + chunk - 1;
	where += chunk;
	bytes -= chunk;
    }
    return 0;
}

static int fini_ith(FILE *f)
{
    /* XXX: entry point? */
    return write_ith_record(f, 0, 0, 1, NULL);
}

static int write_srecord(FILE *f, unsigned int len,  unsigned int alen,
			 uint32_t addr, uint8_t type, void *data)
{
    char buf[2+2+8+255*2+2+2];
    char *p = buf;
    uint8_t csum, *dptr = data;
    unsigned int i;

    if (len > 255) {
	fprintf(stderr, "%s: internal error: invalid srec record size\n",
		progname);
	exit(1);
    }

    switch (alen) {
    case 2:
	addr &= 0xffff;
	break;
    case 3:
	addr &= 0xffffff;
	break;
    case 4:
	break;
    default:
	fprintf(stderr, "%s: internal error: invalid srec address length\n",
		progname);
	exit(1);
    }

    csum = (len+alen+1) + addr + (addr >> 8) + (addr >> 16) + (addr >> 24);
    for (i = 0; i < len; i++)
	csum += dptr[i];
    csum = 0xff-csum;

    p += sprintf(p, "S%c%02X%0*X", type, len+alen+1, alen*2, addr);
    for (i = 0; i < len; i++)
	p += sprintf(p, "%02X", dptr[i]);
    p += sprintf(p, "%02X\n", csum);

    if (fwrite(buf, 1, p-buf, f) != (size_t)(p-buf))
	return -1;

    return 0;
}

static int init_srec(FILE *f)
{
    return write_srecord(f, 0, 2, 0, '0', NULL);
}

static int fini_srec(FILE *f)
{
    /* XXX: entry point? */
    return write_srecord(f, 0, 4, 0, '7', NULL);
}

static int output_srec(FILE *f, void *data, uint32_t bytes, uint32_t where)
{
    uint8_t *dbuf = data;
    unsigned int chunk;

    while (bytes) {
	/* Output up to 32 bytes, but always end on an aligned boundary */
	chunk = 32 - (where & 31);
	if (bytes < chunk)
	    chunk = bytes;

	if (write_srecord(f, chunk, 4, where, '3', dbuf))
	    return -1;

	dbuf += chunk;
	where += chunk;
	bytes -= chunk;
    }
    return 0;
}

static struct output_format output_formats[] = {
    { "bin",  "wb", null_init_fini, output_bin, null_init_fini },
    { "com",  "wb", com_init, output_bin, null_init_fini },
    { "ith",  "wt", null_init_fini, output_ith, fini_ith },
    { "ihx",  "wt", null_init_fini, output_ith, fini_ith },
    { "srec", "wt", init_srec, output_srec, fini_srec },
    { NULL, NULL, NULL, NULL, NULL }
};

static const char *getformat(const char *pathname)
{
    const char *p;
    static char fmt_buf[16];

    /* 
     * Search backwards for the string "rdf2" followed by a string
     * of alphanumeric characters.  This should handle path prefixes,
     * as well as extensions (e.g. C:\FOO\RDF2SREC.EXE).
     */
    for (p = strchr(pathname, '\0')-1 ; p >= pathname ; p--) {
	if (!nasm_stricmp(p, "rdf2")) {
	    const char *q = p+4;
	    char *r = fmt_buf;
	    while (isalnum(*q) && r < fmt_buf+sizeof fmt_buf-1)
		*r++ = *q++;
	    *r = '\0';
	    if (fmt_buf[0])
		return fmt_buf;
	}
     }
    return NULL;
}

static void usage(void)
{
    fprintf(stderr,
	    "Usage: %s [options] input-file output-file\n"
	    "Options:\n"
	    "    -o origin       Specify the relocation origin\n"
	    "    -p alignment    Specify minimum segment alignment\n"
	    "    -f format       Select format (bin, com, ith, srec)\n"
	    "    -q              Run quiet\n"
	    "    -v              Run verbose\n",
	    progname);
}

int main(int argc, char **argv)
{
    rdfmodule *m;
    bool err;
    FILE *of;
    int codepad, datapad;
    const char *format = NULL;
    const struct output_format *fmt;
    bool quiet = false;

    progname = argv[0];

    if (argc < 2) {
	usage();
        return 1;
    }

    argv++, argc--;

    while (argc > 2) {
	if (argv[0][0] == '-' && argv[0][1] && !argv[0][2]) {
	    switch (argv[0][1]) {
	    case 'o':
		argv++, argc--;
		origin = readnum(*argv, &err);
		if (err) {
		    fprintf(stderr, "%s: invalid parameter: %s\n",
			    progname, *argv);
		    return 1;
		}
		origin_def = true;
		break;
	    case 'p':
		argv++, argc--;
		align = readnum(*argv, &err);
		if (err) {
		    fprintf(stderr, "%s: invalid parameter: %s\n",
			    progname, *argv);
		    return 1;
		}
		align_def = true;
		break;
	    case 'f':
		argv++, argc--;
		format = *argv;
		break;
	    case 'q':
		quiet = true;
		break;
	    case 'v':
		quiet = false;
		break;
	    case 'h':
		usage();
		return 0;
	    default:
		fprintf(stderr, "%s: unknown option: %s\n",
			progname, *argv);
		return 1;
	    }
	}
        argv++, argc--;
    }

    if (argc < 2) {
	usage();
        return 1;
    }

    if (!format)
	format = getformat(progname);

    if (!format) {
	fprintf(stderr, "%s: unable to determine desired output format\n",
		progname);
	return 1;
    }

    for (fmt = output_formats; fmt->name; fmt++) {
	if (!nasm_stricmp(format, fmt->name))
	    break;
    }

    if (!fmt->name) {
	fprintf(stderr, "%s: unknown output format: %s\n", progname, format);
	return 1;
    }

    m = rdfload(*argv);

    if (!m) {
        rdfperror(progname, *argv);
        return 1;
    }

    if (!quiet)
	printf("relocating %s: origin=%"PRIx32", align=%d\n",
	       *argv, origin, align);

    m->textrel = origin;
    m->datarel = origin + m->f.seg[0].length;
    if (m->datarel % align != 0) {
        codepad = align - (m->datarel % align);
        m->datarel += codepad;
    } else
        codepad = 0;

    m->bssrel = m->datarel + m->f.seg[1].length;
    if (m->bssrel % align != 0) {
        datapad = align - (m->bssrel % align);
        m->bssrel += datapad;
    } else
        datapad = 0;

    if (!quiet)
	printf("code: %08"PRIx32"\ndata: %08"PRIx32"\nbss:  %08"PRIx32"\n",
	       m->textrel, m->datarel, m->bssrel);

    rdf_relocate(m);

    argv++;

    of = fopen(*argv, fmt->mode);
    if (!of) {
        fprintf(stderr, "%s: could not open output file %s: %s\n",
		progname, *argv, strerror(errno));
        return 1;
    }

    if (fmt->init(of) ||
	fmt->output(of, m->t, m->f.seg[0].length, m->textrel) ||
	fmt->output(of, m->d, m->f.seg[1].length, m->datarel) ||
	fmt->fini(of)) {
        fprintf(stderr, "%s: error writing to %s: %s\n",
		progname, *argv, strerror(errno));
        return 1;
    }

    fclose(of);
    return 0;
}
